Added support for translate

schempy opened the pull request:
I only added changes for translate support. Wanted to make sure I'm on the right track. I'm new to cycle.js and appreciate any feedback/advice. I look forward to contributing! Thanks!

Issue: #2

Widdershin commented:
Hi @schempy, thanks for the pull request!

This is a good start. I wonder if instead of adding a new element helper, if we could instead make `translate` an attribute on each element.

So instead of:

``` js
rect({
  children: [
    translate({x: 50, y: 150}),

    rect({x: 50, y: 50, width: 100, height: 100, draw: [{fill: 'black'}]})
  ]
})
```

Something like:

``` js
rect({
  translate: {x: 50, y: 150},

  children: [
    rect({x: 50, y: 50, width: 100, height: 100, draw: [{fill: 'black'}]})
  ]
})
```

The other thing is that translate is relative, and applies to all future draw operations.

It seems like it would be confusing behavior to have the translate property of one of the elements near the bottom of the tree impacting stuff that isn't actually a child.

To that end, we should probably reset the translate after we finish drawing the element.

What do you think? (If any of this is confusing, please let me know and I'll provide more detailed examples).

Widdershin commented on src/canvas-driver.js:
```diff
@@ -79,9 +79,17 @@ function renderElement (context, element, parent) {
   }
 
+  if (element.kind === 'translate') {
+      performTranslation(context, {x: element.x, y: element.y });
```
Nitpick comments incoming! 😄 

This project uses 2 space indentation, also it looks like there's an extra space just before the closing brace.

schempy commented:
Thanks for the feedback @Widdershin! Everything makes sense. I like the idea of resetting the translate after the draw. I'll be working on the changes.

apretto commented:
If I may, I think it would be better to treat `translate` (and any other transformation) as individual functions, and these functions would then deal with the transformation stack pushing and poping for all its children.

Something like

``` javascript
translate({
  x: 50,
  y: 150,
  children: [
    rect({x: 0, y: 0, width: 100, height: 100, draw: [{fill: 'black'}]}),
    rect({x: 150, y: 0, width: 100, height: 100, draw: [{fill: 'black'}]})
  ]
)
```

would be equivalent to (something like)

``` javascript
ctx.save();
ctx.translate(50, 150);
ctx.fillRect(0, 0, 100, 100);
ctx.fillRect(150, 0, 100, 100);
ctx. restore();
```

I would say that this way composing groups of drawing elements and reusing them, just applying different transformations for different cases. What do you think?

apretto commented:
Giving it more thought, calling `context.save()` and `context.restore()` at the beginning and end of the `renderElement()` function should be safe and work well for all cases, and no special code needs to be added to handle saving and restoring only for transformations (this still assumes an implementation where a transformation is an 'element' and it has a list of children).

I'll try opening a PR for `rotate` for an example.

schempy committed:
Added translate, rotate and scale as attributes for canvas elements.
Also added methods preDrawHooks and postDrawHooks to perform
translations.

schempy commented:
@Widdershin I added the translate, rotate and scale transformations as attributes for elements. I also added `preDrawHooks()` and `postDrawHooks()` to save/restore and perform the transformations.

Let me know what you think.  Thanks!

Widdershin commented:
Looks like we've accidentally duplicated some work, @apretto has added rotation in #5. This is my fault, in that I should really assign issues and make it clear they've been picked up when someone starts working on them. 

Really the decision we have to make here is whether we treat transformations as elements or element attributes. What do y'all think?

@schempy @apretto

apretto commented:
The amount of duplicated work isn't much, so no worries there.

As I said before, I prefer having transformations being elements. I think it'll be better for re-usability of other elements and the transformation itself, and the resulting code would look more functional.

I'll try to make a quick implementation of both styles (transformations as elements vs. as attributes) on the flappy bird demo sometime this weekend, so we can look at the usage to see which approach looks better when in actual use.

schempy commented:
I like transformations as attributes. Makes more sense for me. I'm not opposed to having transformations as elements. Looking forward to @apretto implementations of both styles!

Widdershin commented:
Hey @schempy, @apretto. If you're interested I would love to add y'all as collaborators to this repo, so that we can work together on branches instead of having to make separate pull requests. What do you think? 😄

schempy commented:
Sounds great @Widdershin! Thanks!

Widdershin commented:
Added both of you as contributors 😄.

I'm interested in evolving the API of this library as we run into practical use cases. With that in mind, I think for the sake of moving forward we should merge this PR. I'm still open to the notion of transformations as elements, but that could easily be built on top of as transformations as attributes.

I also think we should move towards getting a v0.1 release so that we can start letting others try this library out. That way, we'll see what users want the most.

Thanks to both of you for putting your time and effort into this 😺.

apretto commented:
Thanks for adding me as a contributor. 👍 

I ran into a problem when trying to update the flappy bird example app to use translate and rotate. The problem is that the way these transformations are performed in a canvas is reminiscent of how these things have been done in computer graphics since the 90s (I would guess). That means there's a hidden state in the canvas with all the transformations done so far. You can see it in the example for rotate in MDN: https://developer.mozilla.org/en-US/docs/Web/API/Canvas_API/Tutorial/Transformations#A_rotate_example

There you'll see that, to rotate the square around the center where it's going to be drawn, you have to:
- translate to the middle of where you'll draw the square (offsets + half the width/height)
- rotate the angle you want
- translate back (just `-1 *` the numbers of the first step)
- draw the square

I dislike this as it creates a cognitive overhead where you have to mentally do the transformations in your head to figure out if they are correct (specially when they are not, as there's a good chance the thing you wanted to draw is going to be off screen and it's harder to figure out which step is the problem).

Some options to approach this problem that I can think of:
1. just expose what the canvas api implements, and let the user of the driver deal with transformations and states;
2. expose the 'low level' api for now, and build abstractions on top of it (for example a `rotateAround` that does that rotate around a specified coordiante)
3. come up with a way to delay applying the transformations, and then process them all in one go so they 'do the right thing'
4. read up and see if anyone figure out a way to do these type of transformations in a functional way.

How it could like for `1`:

``` javascript
group([
  saveState(),
  translate({x: 55, y: 55}),
  rotate({angle: 0.5}),
  translate({x: -55, y: -55}),
  rect({x: 50, y: 50, width: 10, height: 10, draw: [{fill: 'black'}]),
  restoreState()
])
```

How it could like for `2`:

``` javascript
rotateAround(
  {x: 55, y: 55, angle: 0.5},
  [
    rect({x: 50, y: 50, width: 10, height: 10, draw: [{fill: 'black'}])
  ]
)
```

I'm not sure how it could look like for `3` or `4`. One thing that I believe is that these approaches are harder to do having the transformations being attributes to regular draw elements (they would have to be a list to allow the process demonstrated here).

Thoughts?
